### PR TITLE
fix(profit): Filter out NLS from buy-back

### PR DESCRIPTION
**Pull request by KirilMihaylov**

### CODE REVIEW

File: contracts/profit/src/state/buy_back.rs [hunk 1]
```diff
@@ -33,17 +33,27 @@ pub(super) struct BuyBack {
...
-        coins: Vec<CoinDTO<PaymentGroup>>,
-    ) -> Self {
-        Self {
-            contract_addr,
-            config,
-            account,
-            coins,
+        mut coins: Vec<CoinDTO<PaymentGroup>>,
+    ) -> Result<Self, TryNewError> {
+        coins.retain(|coin_dto: &CoinDTO<PaymentGroup>| coin_dto.ticker() != Nls::TICKER);
```
Comment (gmanev7) [inline contracts/profit/src/state/buy_back.rs:42]: define a const for `Nls` or `Nls::Ticker` and use it everywhere in this module

File: contracts/profit/src/state/buy_back.rs [hunk 2]
```diff
@@ -33,17 +33,27 @@ pub(super) struct BuyBack {
 }
 
 impl BuyBack {
-    pub fn new(
+    pub fn try_new(
         contract_addr: Addr,
```
Comment (gmanev7) [inline contracts/profit/src/state/buy_back.rs:37]: [here and the member variable] rename to something more specific than just 'contract', which contract?

**Review by gmanev7 — COMMENTED** [review]
besides the comments here below, there is one moderate issue that smells:
- as partly duplicated code/logic at Idle::send_nls and BuyBack::finish
- as that "ugly" Error -> set data mechanism when there is no balance != Nls at the BuyBack instantiation
my suggestion is to move the 'no-non-NLS-balance' case inside the `BuyBack`

File: contracts/profit/src/state/idle.rs [hunk 3]
```diff
@@ -138,29 +150,93 @@ impl Handler for Idle {
...
+    type Error = Never;
+
+    fn on<C>(self) -> AnyVisitorResult<Self>
+    where
+        C: Currency + Serialize + DeserializeOwned,
+    {
+        Ok(())
+    }
+}
+
```
Comment (gmanev7) [inline contracts/profit/src/state/idle.rs:174]: seems unused

File: contracts/profit/src/state/idle.rs [hunk 4]
```diff
@@ -138,29 +150,93 @@ impl Handler for Idle {
...
+                first: vec![coin.into()],
+                second: Vec::new(),
+            }
+        } else {
+            SplitCoins {
+                first: Vec::new(),
+                second: vec![coin.into()],
+            }
+        })
+    }
```
Comment (gmanev7) [inline contracts/profit/src/state/idle.rs:203]: if it is not in G2?

File: contracts/profit/src/state/idle.rs [hunk 5]
```diff
@@ -43,66 +46,75 @@ impl Idle {
...
     ) -> ContractResult<PlatformResponse>
     where
         B: BankAccount,
     {
-        let state_batch = self.enter(env.block.time, querier)?;
+        let state_response: PlatformResponse =
+            PlatformResponse::messages_only(self.enter(env.block.time, querier)?);
 
-        let balance_nls: Coin<Nls> = account.balance()?;
+        // TODO Generalize to handle all native currencies
```
Comment (gmanev7) [inline contracts/profit/src/state/idle.rs:57]: to alleviate this:
1. revert changes in 'finance::Group'
2. pass 'native' as CoinDTO<Nls>
3. refactor CoinToDTO to split the coins into 'Nls' and 'non-Nls'

File: contracts/profit/src/profit.rs [hunk 6]
```diff
@@ -19,20 +20,23 @@ impl Profit {
         mut account: B,
         env: &Env,
         treasury_addr: &Addr,
-    ) -> ContractResult<(Batch, Emitter)>
+        balance_nls: Coin<Nls>,
```
[the remark this answers is not in the document]
Comment (gmanev7) [reply contracts/profit/src/profit.rs]: not renamed yet

File: contracts/profit/src/state/idle.rs [hunk 7]
```diff
@@ -43,66 +43,76 @@ impl Idle {
         env: &Env,
         querier: &QuerierWrapper<'_>,
         account: B,
+        native: Option<CoinDTO<Native>>,
```
[the remark this answers is not in the document]
Comment (KirilMihaylov) [reply contracts/profit/src/state/idle.rs:46]: I can perform it directly, but it will be hacky.

File: contracts/profit/src/error.rs [hunk 8]
```diff
@@ -34,6 +35,12 @@ pub enum ContractError {
     #[error("[Profit] Alarm comming from unknown address: {0:?}")]
     UnrecognisedAlarm(Addr),
 
+    #[error("[Profit:Buyback] Got currency \"{0}\" when either NLS or one from the buyback group was expected!")]
+    BuybackUnrecognisedCurrency(SymbolStatic),
+
+    #[error("[Profit:Buyback] Broken invariant! {0}")]
+    BuybackBrokenInvariant(String),
+
```
Comment (gmanev7) [inline contracts/profit/src/error.rs:43]: seems unused

File: contracts/profit/src/state/buy_back.rs [hunk 9]
```diff
@@ -24,25 +28,41 @@ use crate::{
...
+        profit_contract: Addr,
         config: Config,
         account: Account,
-        coins: Vec<CoinDTO<BuyBackCurrencies>>,
+        coins: Vec<CoinDTO<PaymentGroup>>,
     ) -> Self {
+        debug_assert!(
+            coins.iter().all(|coin_dto: &CoinDTO<PaymentGroup>| {
+                <
+                currency::non_native_payment::NonNativePaymentGroup as finance::currency::Group
```
Comment (gmanev7) [inline contracts/profit/src/state/buy_back.rs:53]: `NonNativePaymentGroup` is going to be removed ... migrate the code accordingly

Comment (gmanev7) [reply contracts/profit/src/state/buy_back.rs:53]: On a related note, it is not harmful to get some amount in Nls here. There is going to be an extra transfer out to the DEX and that is the only drawback. The SwapExactIn is going to manage its presence and filter it out. Finally, the TransferIn would bring it back to Nolus accumulated into the total amount.

File: contracts/profit/src/state/buy_back.rs [hunk 10]
```diff
@@ -157,3 +176,29 @@ where
...
+        type Error = Never;
+
+        fn on<C>(self) -> AnyVisitorResult<Self>
+        where
+            C: Currency + Serialize + DeserializeOwned,
+        {
+            Ok(())
+        }
+    }
+}
```
Comment (gmanev7) [inline contracts/profit/src/state/buy_back.rs:204]: it looks like an overkill solution ... and non-applicable since we do not have Payment-NLS group

**Review by gmanev7 — APPROVED** [review]

**Pull request merged**